In [1]:
import torch

# Задание 1: Создание и манипуляции с тензорами (25 баллов)


## 1.1 Создание тензоров (7 баллов)


In [4]:
# Установка сида для воспроизводимости
torch.manual_seed(46)

# Создайте следующие тензоры:
# - Тензор размером 3x4, заполненный случайными числами от 0 до 1
torch.manual_seed(46)
tensor3x4_rand= torch.rand(3, 4)
# - Тензор размером 2x3x4, заполненный нулями
tensor2x3x4_zeros = torch.zeros((2, 3, 4))
# - Тензор размером 5x5, заполненный единицами
tensor5x5_ones = torch.ones(5,5)
# - Тензор размером 4x4 с числами от 0 до 15 (используйте reshape)
tensor4x4 = torch.arange(0, 16)
tensor4x4 = tensor4x4.reshape((4,4))

print('Случайный:\n',tensor3x4_rand)
print('Нулевой:\n', tensor2x3x4_zeros)
print('Единичный:\n', tensor5x5_ones)
print('Последовательный:\n', tensor4x4)

Случайный:
 tensor([[0.6611, 0.0600, 0.5174, 0.1596],
        [0.7550, 0.8390, 0.0674, 0.4631],
        [0.1477, 0.3597, 0.9328, 0.0170]])
Нулевой:
 tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Единичный:
 tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])
Последовательный:
 tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15]])


## 1.2 Операции с тензорами (6 баллов)


In [6]:
# Установка сида для воспроизводимости
torch.manual_seed(46)

# Дано: тензор A размером 3x4 и тензор B размером 4x3
# Выполните:
# - Транспонирование тензора A
# - Матричное умножение A и B
# - Поэлементное умножение A и транспонированного B
# - Вычислите сумму всех элементов тензора A

A = torch.rand(3, 4)
B = torch.rand(4, 3)

T_tensor = A.T
matmul_tensor = A@B
mul_A_BTransform = A*B.T
sum_A_elems = A.sum()

print('Транспонированный тензор A:\n', T_tensor)
print('Матричное умножение A на B:\n ', matmul_tensor)
print('Поэлементное умножение A и транспонированного B:\n', mul_A_BTransform)
print('Сумма всех элементов тензора A:\n', sum_A_elems)

Транспонированный тензор A:
 tensor([[0.6611, 0.7550, 0.1477],
        [0.0600, 0.8390, 0.3597],
        [0.5174, 0.0674, 0.9328],
        [0.1596, 0.4631, 0.0170]])
Матричное умножение A на B:
  tensor([[1.0578, 0.8108, 0.6521],
        [1.8322, 1.2522, 1.3869],
        [0.9210, 0.9986, 0.4515]])
Поэлементное умножение A и транспонированного B:
 tensor([[0.6436, 0.0528, 0.2489, 0.1125],
        [0.3102, 0.5512, 0.0498, 0.3411],
        [0.1273, 0.2932, 0.0291, 0.0018]])
Сумма всех элементов тензора A:
 tensor(4.9799)


## 1.3 Индексация и срезы (6 баллов)


In [8]:
# Установка сида для воспроизводимости
torch.manual_seed(46)

# Создайте тензор размером 5x5x5
tensor5x5x5 = torch.rand((5,5,5))

# Извлеките:
# - Первую строку
print('Первая строка:\n', tensor5x5x5[:, 0, :])

# - Последний столбец
print('Последний столбец:\n', tensor5x5x5[:, :, -1])

# - Подматрицу размером 2x2 из центра тензора#
# Для каждых матриц в оси 0 вычисляем центр 
print('Центры 2х2:\n', tensor5x5x5[:, 1:3, 1:3])


# - Все элементы с четными индексами
print('С четными индексами:\n', tensor5x5x5[::2, ::2, ::2])

Первая строка:
 tensor([[0.6611, 0.0600, 0.5174, 0.1596, 0.7550],
        [0.2633, 0.9035, 0.6618, 0.9728, 0.9471],
        [0.5747, 0.9875, 0.1572, 0.1534, 0.7301],
        [0.0212, 0.1792, 0.0934, 0.6219, 0.9400],
        [0.5990, 0.6566, 0.3949, 0.1408, 0.5100]])
Последний столбец:
 tensor([[0.7550, 0.3597, 0.8620, 0.7388, 0.1455],
        [0.9471, 0.3397, 0.5063, 0.1992, 0.0602],
        [0.7301, 0.7769, 0.1473, 0.9561, 0.1412],
        [0.9400, 0.8611, 0.9862, 0.9526, 0.1365],
        [0.5100, 0.9711, 0.5039, 0.5731, 0.5805]])
Центры 2х2:
 tensor([[[0.0674, 0.4631],
         [0.0170, 0.9736]],

        [[0.9694, 0.6430],
         [0.0770, 0.1563]],

        [[0.6019, 0.7746],
         [0.4427, 0.1632]],

        [[0.8525, 0.3752],
         [0.4504, 0.6682]],

        [[0.4292, 0.6907],
         [0.3050, 0.9780]]])
С четными индексами:
 tensor([[[0.6611, 0.5174, 0.7550],
         [0.9328, 0.9736, 0.8620],
         [0.0312, 0.7364, 0.1455]],

        [[0.5747, 0.1572, 0.7301],
     

## 1.4 Работа с формами (6 баллов)


In [10]:
# Установка сида для воспроизводимости
torch.manual_seed(46)

# Создайте тензор размером 24 элемента
string_tensor = torch.arange(24)
# Преобразуйте его в формы:
# - 2x12
shape2x12 = string_tensor.reshape(2, 12)

# - 3x8
shape3x8 = string_tensor.reshape(3, 8)

# - 4x6
shape4x6 = string_tensor.reshape(4,6)

# - 2x3x4
shape2x3x4 = string_tensor.reshape((2,3,4))

# - 2x2x2x3
shape2x2x2x3 = string_tensor.reshape((2,2,2,3))

print('2x12:\n', shape2x12)
print('3x8:\n', shape3x8)
print('4x6:\n', shape4x6)
print('2x3x4:\n', shape2x3x4)
print('2x2x2x3:\n', shape2x2x2x3)

2x12:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])
3x8:
 tensor([[ 0,  1,  2,  3,  4,  5,  6,  7],
        [ 8,  9, 10, 11, 12, 13, 14, 15],
        [16, 17, 18, 19, 20, 21, 22, 23]])
4x6:
 tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])
2x3x4:
 tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])
2x2x2x3:
 tensor([[[[ 0,  1,  2],
          [ 3,  4,  5]],

         [[ 6,  7,  8],
          [ 9, 10, 11]]],


        [[[12, 13, 14],
          [15, 16, 17]],

         [[18, 19, 20],
          [21, 22, 23]]]])


# Задание 2: Автоматическое дифференцирование (25 баллов)


## 2.1 Простые вычисления с градиентами (8 баллов)


In [13]:
# Установка сида для воспроизводимости
torch.manual_seed(46)

# Создайте тензоры x, y, z с requires_grad=True
x = torch.rand(6, requires_grad=True)
y = torch.rand(6, requires_grad=True)
z = torch.rand(6, requires_grad=True)
print(f'x: {x}')
print(f'y: {y}')
print(f'z: {z}')
print()

# Вычислите функцию: f(x,y,z) = x^2 + y^2 + z^2 + 2*x*y*z
f = x**2 + y**2 + z**2 + 2*x*y*z
# Обратное распространение (Auto grad) 
f.sum().backward()

# Найдите градиенты по всем переменным
print('Градиент x:', x.grad)
print('Градиент y:', y.grad)
print('Градиент z:', z.grad)

x: tensor([0.6611, 0.0600, 0.5174, 0.1596, 0.7550, 0.8390], requires_grad=True)
y: tensor([0.0674, 0.4631, 0.1477, 0.3597, 0.9328, 0.0170], requires_grad=True)
z: tensor([0.9736, 0.4108, 0.8620, 0.8799, 0.6569, 0.8152], requires_grad=True)

Градиент x: tensor([1.4534, 0.5005, 1.2895, 0.9523, 2.7355, 1.7058])
Градиент y: tensor([1.4220, 0.9755, 1.1874, 1.0003, 2.8575, 1.4020])
Градиент z: tensor([2.0363, 0.8772, 1.8769, 1.8747, 2.7223, 1.6590])


Для вычисления градиента нужно подсчитать производную функции f для каждого из тензоров:
- **df(x,y,z)/dx** = 2x + 2yz 
- **df(x,y,z)/dy** = 2y + 2xz 
- **df(x,y,z)/dz** = 2z + 2xy 

После нужно подставлять значения кадого элемента из тензоров. Например (для torch.manual_seed(46)):\
>
> x[0] = 0.6611. y[0] = 0.0674, z[0] = 0.9736\
> **df(x,y,z)/dx** = 2 * 0.6611 + 2 * 0.0674 * 0.9736 = **1.4534**4128\
>Это совпадает с полученным результатом x.grad[0] = 1.4534

## 2.2 Градиент функции потерь (9 баллов)


In [16]:
# Реализуйте функцию MSE (Mean Squared Error):
def MSE(y_pred, y_true):
    # MSE = (1/n) * Σ(y_pred - y_true)^2
    return torch.mean((y_pred - y_true) ** 2)

# где y_pred = w * x + b (линейная функция)
# Найдите градиенты по w и b


# Истинные значения, которые должна выдавать линейная модель
true_values = torch.tensor([2, 4, 6, 8])
# Входные данные
raw_values = torch.tensor([1, 2, 3, 4])

# Назначим условные значения для коэффициента и смещения
w = torch.tensor(1.47, requires_grad=True)
b = torch.tensor(0.14, requires_grad=True)

# Предзказание линейной регрессии
y_pred = w * raw_values + b

# Вычисление MSE и обратное расспространение
mse = MSE(y_pred, true_values)
mse.backward()

print('Градиент коэффициента: ', w.grad)
print('Градиент смещения: ', b.grad)

Градиент коэффициента:  tensor(-7.2500)
Градиент смещения:  tensor(-2.3700)


## 2.3 Цепное правило (8 баллов)


In [18]:
# Реализуйте составную функцию: f(x) = sin(x^2 + 1)
def F_sin(x):
    return torch.sin(x**2 + 1)

# Какое-нибуль значение
value = torch.tensor([4.0], requires_grad=True)

### Найдите градиент df/dx
Найдем производную по x (то есть в данном случае градиент):
>f`(x) = 2x\*cos(x**2 + 1)
>
>f`(4) = 2\*4 \* cos(16 + 1) = 8cos(17) = 8 \* −0,27516333805 = −2,2013067044

Получили ответ: **−2,2013**067044

In [20]:
# Проверьте результат с помощью torch.autograd.grad
answer = F_sin(value)
print('Градиент: ', torch.autograd.grad(answer, value))

Градиент:  (tensor([-2.2013]),)


# Задание 3: Сравнение производительности CPU vs CUDA (20 баллов)


## 3.1 Подготовка данных (5 баллов)


In [23]:
# Создайте большие матрицы размеров:
# - 64 x 1024 x 1024
big64x1024x1024 = torch.rand((64, 1024, 1024))

# - 128 x 512 x 512
big128x512x512 = torch.rand((128,512,512))

# - 256 x 256 x 256
big256x256x256 = torch.rand((256,256,256))

# Заполните их случайными числами (сделано)

## 3.2 Функция измерения времени (5 баллов)


In [25]:
# Создайте функцию для измерения времени выполнения операций
# Используйте torch.cuda.Event() для точного измерения на GPU
# Используйте time.time() для измерения на CPU
import time

def GPU(tensor, operator):
    tensor = tensor.to(torch.device('cuda'))
    # Инициализируем cuda Events и активируем счетчик времени
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    # Для усреднения результата проводим вычисления 10 раз
    for i in range(15):
        operator(tensor)
    end.record()
    torch.cuda.synchronize()  # Ждем пока все ядра закончат работать  
    return start.elapsed_time(end) / 15

def CPU(tensor, operator):
    start = time.time()
    # Для усреднения результата проводим вычисления 10 раз
    for i in range(10):
        operator(tensor)
    total_time = time.time() - start
    return total_time / 15


## 3.3 Сравнение операций (10 баллов)


In [27]:
# Сравните время выполнения следующих операций на CPU и CUDA:
# - Матричное умножение (torch.matmul)
def matmul(tensor):
    return torch.matmul(tensor, tensor)
    
# - Поэлементное сложение
def add(tensor):
    return tensor + tensor

# - Поэлементное умножение
def multiply(tensor):
    return tensor * tensor
  
# - Транспонирование
def transposition(tensor):
    return tensor.T
    
# - Вычисление суммы всех элементов
def summurize(tensor):
    return tensor.sum()

# Для каждой операции:
# 1. Измерьте время на CPU
matmulCPU = CPU(big64x1024x1024, matmul) * 1000
addCPU = CPU(big64x1024x1024, add) * 1000
multiplyCPU = CPU(big64x1024x1024, multiply) * 1000
transpositionCPU = CPU(big64x1024x1024, transposition) * 1000
summurizeCPU = CPU(big64x1024x1024, summurize) * 1000

# 2. Измерьте время на GPU (если доступен)
matmulGPU = GPU(big64x1024x1024, matmul)
addGPU = GPU(big64x1024x1024, add)
multiplyGPU = GPU(big64x1024x1024, multiply)
transpositionGPU = GPU(big64x1024x1024, transposition)
summurizeGPU = GPU(big64x1024x1024, summurize)

C:\Users\bigbl\AppData\Local\Temp\ipykernel_13416\2909483635.py:16: UserWarning: The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated and it will throw an error in a future release. Consider `x.mT` to transpose batches of matrices or `x.permute(*torch.arange(x.ndim - 1, -1, -1))` to reverse the dimensions of a tensor. (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\pytorch\aten\src\ATen\native\TensorShape.cpp:4416.)
  return tensor.T


In [28]:
# 3. Вычислите ускорение (speedup)
matmulSpeedUp =  round(matmulCPU / matmulGPU, 5)
addSpeedUP =  round(addCPU/addGPU, 5)
multiplySpeedUp =  round(multiplyCPU/multiplyGPU, 5)
# transpositionSpeedUp = transpositionGPU / transpositionCPU
summurizeSpeedUp =  round(summurizeCPU/summurizeGPU, 5)

# 4. Выведите результаты в табличном виде
print('Операция\t\t| CPU(мс)\t     | GPU(мс) \t\t | Ускорение')
print(f'Матричное умножение\t| {matmulCPU}  | {matmulGPU} | {matmulSpeedUp}')
print(f'Поэлементное умножение\t| {multiplyCPU} | {multiplyGPU}| {multiplySpeedUp}')
print(f'Поэлементное сложение\t| {addCPU} | {addCPU}| {addSpeedUP}')
print(f'Транспонирование\t| {transpositionCPU}| {round(transpositionGPU, 5)}\t\t | ?')
print(f'Вычисление суммы\t| {summurizeCPU}  | {summurizeGPU}| {summurizeSpeedUp}')


Операция		| CPU(мс)	     | GPU(мс) 		 | Ускорение
Матричное умножение	| 298.560889561971  | 41.857906087239584 | 7.13272
Поэлементное умножение	| 29.961935679117836 | 3.4706944783528644| 8.63284
Поэлементное сложение	| 29.361168543497723 | 29.361168543497723| 7.89456
Транспонирование	| 0.0| 0.00632		 | ?
Вычисление суммы	| 7.420269648234049  | 2.1007872263590497| 3.53214


## 3.4 Анализ результатов (5 баллов)


Проанализируйте результаты:
- Какие операции получают наибольшее ускорение на GPU?
  >Матричное умножение, поэлементное сложение и поэлементное умножение, так как они являются самыми затратными и GPU пожет расспараллелить эти вычисления
- Почему некоторые операции могут быть медленнее на GPU?
  > Это связано с легкостью вычисления, например, транспонирования: GPU трати время на передачу данных с CPU на GPU. Также есть свои накладные алгоритмы у GPU, поэтому вычисление легких задач на CPU быстрее
- Как размер матриц влияет на ускорение?
- Что происходит при передаче данных между CPU и GPU?